pennylane/qnodes/passthru.py

```diff
@@ -0,0 +1,145 @@
+    autodiff (AD) framework (such as TensorFlow or PyTorch). This means that the QNode
+    converts all its inputs (which may come in data types specific to the
+    AD framework used, which we denote *ADT* here) into plain Python/NumPy types,
+    computes the required :ref:`quantum function <intro_vcirc_qfunc>` value or Jacobian,
+    and converts the result back into the ADT if necessary.
+
+    In contrast, PassthruQNode works as a white box: it preserves the ADT
+    throughout the computation. This requires that the quantum function is computed
+    using a simulator device that is compatible with the AD framework used (typically
+    implemented using that same framework).
+
+    The advantages of this approach is that the qfunc can be differentiated using its AD framework
+    without requiring a separate method for computing the Jacobian, and that the internals
+    of the simulation are visible in the computational graph.
+
+    PassthruQNode requires some changes to the way other PennyLane components work:
+
+    1. :class:`Operator` must not do domain checking for its parameters, or it must let the ADT pass the check.
+    2. The simulator device must return the result as the ADT instead of plain Python/NumPy types.
+    3. Any output_conversion in :meth:`BaseQNode.evaluate` must be skipped.
+
+    Additionally, any array-like ADT needs to be able to handle (1) scalar multiplication,
+    (2) indexing/slicing, and possibly (3) iteration, as these are the things qfuncs expect of
+    array-like parameters.
+    """
+    def __init__(self, func, device, properties=None):
+        super().__init__(func, device, mutable=True, properties=properties)
+
+    def __repr__(self):
+        """String representation."""
+        detail = "<PassthruQNode: device='{}', func={}, wires={}>"
+        return detail.format(self.device.short_name, self.func.__name__, self.num_wires)
+
+    def _set_variables(self, args, kwargs):
+        # do nothing, since we do not use VariableRefs
+        pass
+
+    def _construct(self, args, kwargs):
+        """Construct the quantum circuit graph by calling the quantum function.
+
+        Like :class:`.BaseQNode._construct`, but does not use VariableRefs.
+
+        PassthruQNode does not have a Jacobian method, so it does not HAVE to use VariableRefs or scalar linear indexing of input parameters.
+        Two options:
+        1. Use VariableRefs anyway, re-use most BaseQNode methods.
+           Problem: after evaluating the VariableRefs, stacking sliced/indexed Tensors in Operation.parameters should somehow result in a Tensor, not an object array.
+        2. Do not use VariableRefs, call the qfunc each time :meth:`PassthruQNode.evaluate` is called (always mutable).
+           Problem: tensornet_tf requires variable_deps?
+
+
+        TODO rethink output_conversion? should require device to return things in a fixed form, but either as arrays or as AD Tensors, do conversion in interface (if necessary...)
```

Comment:
I think this whole interface and output conversion logic should be reviewed and made coherent. A long time ago I thought about abstracting the whole logic and to basically beef up the interfaces to also handle conversion of inputs and outputs -- maybe this would be sensible?

Reply:
I'm almost always in favor of consistent interfaces, even if they are a bit more unwieldy. In this case the main question is whether a qfunc returning just a single scalar should return a `float` or an `array[float]` of shape `(1,)`.  The former is maybe easier for the user (no need to use indexing, `result[0]`, to get at the number), the latter is more consistent (qfunc always returns a 1-d array, you can always e.g. iterate through the result).

Also note that `PassthruQNode` does not use the `pennylane.interfaces`. The interfaces are necessary to make black box QNodes interact with AD frameworks, whereas `PassthruQNode` is a white box, completely transparent to the AD framework.

Reply:
I definitely think the library is due for a revisit/relook/refactor to how the interfaces work, now that we have more experience working with them and developing against them. Should be in scope for a future milestone, but good to start thinking about it now

Reply:
I think the main takeaway here is a revisit of the `QNode.evaluate` post-processing logic, rather than the interface logic. `QNode.evaluate` is where the output conversion @smite mentions currently occurs, which results in arrays of length 1 treated differently from arrays of length > 1.

Note that the current interfaces don't do any post-processing (apart from changing the type from numpy -> Tensor where required), so in a sense are already kind of 'consistent'. We could move this output conversion to the interfaces, or we could drop it altogether.

Reply:
I can't quite remember the reason for the output conversion, but if recall, it goes down to the Device - `Device.execute` will always return an array, and so we need to index/use `squeeze` to remove excessive dimensions.

Reply:
What I imagined was to have some kind of plug-and-play logic for the conversions that derives from an abstract baseclass that provides methods like `to_array(input), to_scalar(input)` or so. One could have those conversions for all sorts of inputs and also have a `PassthruWrapper` that just does nothing.

Reply:
Nice suggestion @johannesjmeyer

Reply:
I think it'd be good to consider if the output conversions are even necessary. They make certain use cases slightly more convenient, but this comes at the cost of a lack of consistency, and a bit of extra code of course.

Reply:
If I recall, there were two reasons we had for having the output conversion:

1. UI: A QNode that returns a single output should return a scalar, not an array with shape `[1]` (mostly just a UI issue). Meanwhile, a QNode with a tuple/list/sequence of measurements should return output as a sequence.

2. Autograd issues: We could not get Autograd working if the output was a sequence, we had to enforce conversion to a NumPy array.

But it's been so long, I'm in favour of double checking and re-exploring.

Reply:
And we also had 
3. QNodes can return expvals, vars and samples along-side each other, which necessitates the use of `np.squeeze`.
And I guess due to the fact that autograd isn't actively developed anymore that issue 2. didn't change in the meantime.